src/GraphQL.DataLoader/DataLoaderPair.cs

```diff
@@ -30,17 +30,29 @@ public DataLoaderPair(IDataLoader loader, TKey key)
         /// <summary>
         /// Returns the data loader that is called when the result is requested
         /// </summary>
-        public IDataLoader Loader { get; }
+        private IDataLoader? _loader;
 
         /// <summary>
         /// Returns the result if it has been set, or throws an exception if not
         /// </summary>
-        public T Result => IsResultSet ? _result : throw new InvalidOperationException("Result has not been set");
+        public T Result
+        {
+            get
+            {
+                if (IsResultSet)
+                {
+                    Interlocked.MemoryBarrier(); // ensure that _loader is read before _result is read
```

Comment:
🤔 A rather rare API to see, I don’t even have experience working with it. What is its help here?

Reply:
It’s use case is also very rare. Basically only is useful in situations like these where a lock is not necessary so long as the Intel CPU always reads from and writes to memory in the order stated. (Typically the compiler and even the CPU itself can reorder memory reads during its optimization)

Reply:
The code comment explains precisely why the line of code was added; hopefully it is clear.